File: 155-Min-Stack/solution.cpp
class MinStack {
private:
    stack<int> myStack, minStack;
public:
    void push(int x) {
        if(myStack.empty() || minStack.empty()) {
            myStack.push(x);
            minStack.push(x);
        } else {
            myStack.push(x);
            minStack.push((x < minStack.top()) ? x : minStack.top());
            
        }
    }

    void pop() {
        if(!myStack.empty() && !minStack.empty()) {
            myStack.pop();
            minStack.pop();
        }
    }

    int top() {
        return myStack.top();
    }

    int getMin() {
        return minStack.top();
    }
};

File: 155-Min-Stack/solution.java
class MinStack {

    private Stack<Integer> stack, min;

    /** initialize your data structure here. */
    public MinStack() {
        this.stack = new Stack();
        this.min = new Stack();
    }

    public void push(int x) {
        stack.push(x);
        if (min.isEmpty() || min.peek() >= x) {
            min.push(x);
        } else {
            min.push(min.peek());
        }
    }

    public void pop() {
        min.pop();
        stack.pop();
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return min.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */